temporal-sdk/src/main/java/io/temporal/internal/worker/WorkflowRunLockManager.java

```diff
@@ -19,89 +19,101 @@
 
```

Comment:
I think the approach from https://stackoverflow.com/a/23920910 is easier to follow. `ConcurrentHashMap` + `BlockingQueue` since the latter has `offer` and `poll` with built in timeouts.

Reply:
The approach from this link will not work because it will leak. If the approach by this link would be acceptable, I would just make a map from keys to lock and be done with it.